File: src/main/java/me/miki/shindo/ui/comp/factory/SettingComponentFactory.java
package me.miki.shindo.ui.comp.factory;

import me.miki.shindo.management.settings.Setting;
import me.miki.shindo.management.settings.impl.BooleanSetting;
import me.miki.shindo.management.settings.impl.CellGridSetting;
import me.miki.shindo.management.settings.impl.ColorSetting;
import me.miki.shindo.management.settings.impl.ComboSetting;
import me.miki.shindo.management.settings.impl.ImageSetting;
import me.miki.shindo.management.settings.impl.KeybindSetting;
import me.miki.shindo.management.settings.impl.NumberSetting;
import me.miki.shindo.management.settings.impl.SoundSetting;
import me.miki.shindo.management.settings.impl.TextSetting;
import me.miki.shindo.ui.comp.Comp;
import me.miki.shindo.ui.comp.impl.CompCellGrid;
import me.miki.shindo.ui.comp.impl.CompColorPicker;
import me.miki.shindo.ui.comp.impl.CompComboBox;
import me.miki.shindo.ui.comp.impl.CompImageSelect;
import me.miki.shindo.ui.comp.impl.CompKeybind;
import me.miki.shindo.ui.comp.impl.CompModTextBox;
import me.miki.shindo.ui.comp.impl.CompSlider;
import me.miki.shindo.ui.comp.impl.CompSoundSelect;
import me.miki.shindo.ui.comp.impl.CompToggleButton;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.function.Function;

/**
 * Central registry responsible for turning {@link Setting} instances into UI components. This makes
 * it easy to plug new component types without editing the panel code and keeps the factory logic in one place.
 */
public final class SettingComponentFactory {

    private static final Map<Class<? extends Setting>, Function<Setting, Comp>> REGISTRY = new LinkedHashMap<>();

    static {
        register(BooleanSetting.class, setting -> new CompToggleButton((BooleanSetting) setting));
        register(NumberSetting.class, setting -> new CompSlider((NumberSetting) setting));
        register(ComboSetting.class, setting -> new CompComboBox(140, (ComboSetting) setting));
        register(KeybindSetting.class, setting -> new CompKeybind(120, (KeybindSetting) setting));
        register(TextSetting.class, setting -> new CompModTextBox((TextSetting) setting));
        register(ColorSetting.class, setting -> new CompColorPicker((ColorSetting) setting));
        register(ImageSetting.class, setting -> new CompImageSelect((ImageSetting) setting));
        register(SoundSetting.class, setting -> new CompSoundSelect((SoundSetting) setting));
        register(CellGridSetting.class, setting -> new CompCellGrid(270, 160, (CellGridSetting) setting));
    }

    private SettingComponentFactory() {
    }

    public static void register(Class<? extends Setting> type, Function<? super Setting, ? extends Comp> factory) {
        REGISTRY.put(type, setting -> factory.apply(setting));
    }

    public static Comp create(Setting setting) {
        for (Map.Entry<Class<? extends Setting>, Function<Setting, Comp>> entry : REGISTRY.entrySet()) {
            if (entry.getKey().isInstance(setting)) {
                return entry.getValue().apply(setting);
            }
        }
        return null;
    }
}

File: src/main/kotlin/me/miki/shindo/ui/comp/factory/SettingComponentFactory.kt
package me.miki.shindo.ui.comp.factory

import me.miki.shindo.management.settings.Setting
import me.miki.shindo.management.settings.impl.*
import me.miki.shindo.ui.comp.Comp
import me.miki.shindo.ui.comp.buttons.CompToggleButton
import me.miki.shindo.ui.comp.inputs.*
import java.util.concurrent.ConcurrentHashMap
object SettingComponentFactory {
    private val registry: MutableMap<Class<out Setting>, (Setting) -> Comp> = ConcurrentHashMap()
    private val componentCache: MutableMap<Setting, Comp> = ConcurrentHashMap()

    init {
        registerDefaultFactories()
    }

    private fun registerDefaultFactories() {
        register(BooleanSetting::class.java) { setting ->
            CompToggleButton(setting)
        }
        register(NumberSetting::class.java) { setting ->
            CompSlider(setting)
        }
        register(ComboSetting::class.java) { setting ->
            CompComboBox(140f, setting)
        }
        register(KeybindSetting::class.java) { setting ->
            CompKeybind(120f, setting)
        }
        register(TextSetting::class.java) { setting ->
            CompModTextBox(setting)
        }
        register(ColorSetting::class.java) { setting ->
            CompColorPicker(setting)
        }
        register(ImageSetting::class.java) { setting ->
            CompImageSelect(setting)
        }
        register(SoundSetting::class.java) { setting ->
            CompSoundSelect(setting)
        }
        register(CellGridSetting::class.java) { setting ->
            CompCellGrid(270f, 160f, setting)
        }
    }
    @JvmStatic
    fun <T : Setting> register(type: Class<T>, factory: (T) -> Comp) {
        @Suppress("UNCHECKED_CAST")
        registry[type] = factory as (Setting) -> Comp
    }
    @JvmStatic
    fun create(setting: Setting): Comp? {

        componentCache[setting]?.let { return it }

        val settingClass = setting.javaClass
        val factory = findFactory(settingClass)
                ?: return null

        val component = factory(setting)
        componentCache[setting] = component
        return component
    }
    private fun findFactory(settingClass: Class<out Setting>): ((Setting) -> Comp)? {

        registry[settingClass]?.let { return it }

        var current: Class<*>? = settingClass
        while (current != null && Setting::class.java.isAssignableFrom(current)) {
            @Suppress("UNCHECKED_CAST")
            registry[current as Class<out Setting>]?.let { return it }
            current = current.superclass
        }

        return null
    }
    @JvmStatic
    fun clearCache() {
        componentCache.clear()
    }
    @JvmStatic
    fun invalidateCache(setting: Setting) {
        componentCache.remove(setting)
    }
    @JvmStatic
    fun getRegisteredFactoryCount(): Int = registry.size
}